tests/unit/steps/test_steps_plan.py

```diff
@@ -920,5 +1025,44 @@ async def test_generate_plan_upgrade_group_data_plane(
     mock_determine_upgrade_target.assert_called_once()
     mock_pre_upgrade_steps.assert_called_once()
     mock_control_plane.assert_not_called()
-    mock_data_plane.assert_called_once()
+    mock_separate_hypervisors_apps.assert_called_once()
+
+    mock_generate_hypervisors_plan.assert_called_once()
+    mock_ceph_osd_subordinates.assert_not_called()
     mock_post_upgrade_steps.assert_called_once()
+
+
+def test_separate_hypervisors_apps():
+    nova_compute = MagicMock(spec_set=OpenStackApplication)()
+    nova_compute.charm = "nova-compute"
+    nova_compute.is_subordinate = False
+
+    cinder = MagicMock(spec_set=OpenStackApplication)()
+    cinder.charm = "cinder"
+    cinder.is_subordinate = False
+
+    ceph_osd = MagicMock(spec_set=OpenStackApplication)()
+    ceph_osd.charm = "ceph-osd"
+    ceph_osd.is_subordinate = False
+
+    ovn_chassis = MagicMock(spec_set=OvnSubordinate)()
+    ovn_chassis.charm = "ovn-chassis"
+    ovn_chassis.is_subordinate = True
+
+    result = cou_plan._separate_hypervisors_apps([nova_compute, cinder, ceph_osd, ovn_chassis])
+    assert result == ([nova_compute, cinder], [ceph_osd, ovn_chassis])
+
+
+@patch("cou.steps.plan.create_upgrade_group")
+def test_generate_ceph_osd_and_subordinates_plan(mock_create_upgrade_group):
+    target = OpenStackRelease("victoria")
+    non_hypervisors_apps = [
+        MagicMock(spec_set=OpenStackApplication)(),
+        MagicMock(spec_set=OvnSubordinate)(),
+    ]
+    result = cou_plan._generate_ceph_osd_and_subordinates_plan(target, non_hypervisors_apps, False)
+    assert mock_create_upgrade_group.call_count == 2
```

Comment:
We should check calls itself and not only checking that there were 2 calls. So we need to use `assert_has_calls` and define those calls.

Reply:
I agree...
The problem was that we were passing lambda functions as arguments for `create_upgrade_group` and it's very hard to test it. If it's hard to test, it means that the code is not good, so I've changed a little bit in a way that now we can check what was called

Reply:
I do not understand why it's hard to test it. Can you provide some example?

Reply:
You can check the it with [`ANY`](https://docs.python.org/3/library/unittest.mock.html#any) is not perfect, but fine. I agree that the code is not the best. I really think we should avoid `create_upgrade_group`.